// @luma.gl/core/dist/index.js
// luma.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
export { luma } from "./adapter/luma.js";
// ADAPTER (DEVICE AND GPU RESOURCE INTERFACES)
export { Adapter } from "./adapter/adapter.js";
export { Device, DeviceFeatures, DeviceLimits } from "./adapter/device.js";
export { CanvasContext } from "./adapter/canvas-context.js";
export { Resource } from "./adapter/resources/resource.js";
export { Buffer } from "./adapter/resources/buffer.js";
export { Texture } from "./adapter/resources/texture.js";
export { TextureView } from "./adapter/resources/texture-view.js";
export { ExternalTexture } from "./adapter/resources/external-texture.js";
export { Shader } from "./adapter/resources/shader.js";
export { Sampler } from "./adapter/resources/sampler.js";
export { Framebuffer } from "./adapter/resources/framebuffer.js";
export { RenderPipeline } from "./adapter/resources/render-pipeline.js";
export { RenderPass } from "./adapter/resources/render-pass.js";
export { ComputePipeline } from "./adapter/resources/compute-pipeline.js";
export { ComputePass } from "./adapter/resources/compute-pass.js";
export { CommandEncoder } from "./adapter/resources/command-encoder.js";
export { CommandBuffer } from "./adapter/resources/command-buffer.js";
export { VertexArray } from "./adapter/resources/vertex-array.js";
export { TransformFeedback } from "./adapter/resources/transform-feedback.js";
export { QuerySet } from "./adapter/resources/query-set.js";
export { PipelineLayout } from "./adapter/resources/pipeline-layout.js";
// PORTABLE API - UNIFORM BUFFERS
export { UniformBufferLayout } from "./portable/uniform-buffer-layout.js";
export { UniformBlock } from "./portable/uniform-block.js";
export { UniformStore } from "./portable/uniform-store.js";
// GPU TYPE UTILS - GPU MEMORY LAYOUT HELPERS - CAN BE USED BY APPS BUT MOSTLY USED INTERNALLY
export { decodeVertexFormat } from "./gpu-type-utils/decode-vertex-format.js";
export { decodeShaderUniformType } from "./gpu-type-utils/decode-shader-types.js";
export { decodeShaderAttributeType } from "./gpu-type-utils/decode-attribute-type.js";
export { getDataTypeFromTypedArray } from "./gpu-type-utils/vertex-format-from-attribute.js";
export { getTypedArrayFromDataType } from "./gpu-type-utils/vertex-format-from-attribute.js";
export { getVertexFormatFromAttribute } from "./gpu-type-utils/vertex-format-from-attribute.js";
export { decodeTextureFormat } from "./gpu-type-utils/decode-texture-format.js";
export { getTextureFormatCapabilities } from "./gpu-type-utils/texture-format-capabilities.js";
// INTERNAL UTILS - for use in other luma.gl modules only
export { log } from "./utils/log.js";
export { getScratchArray } from "./utils/array-utils-flat.js";
export { BufferLayoutHelper as _BufferLayoutHelper } from "./adapter-utils/buffer-layout-helper.js";
export { getAttributeInfosFromLayouts } from "./adapter-utils/get-attribute-from-layouts.js";
export { sortedBufferLayoutByShaderSourceLocations } from "./adapter-utils/buffer-layout-order.js";
// TEST EXPORTS
export { getTextureFormatDefinition as _getTextureFormatDefinition, getTextureFormatTable as _getTextureFormatTable } from "./gpu-type-utils/texture-format-table.js";
